forge/src/trace/decoder.rs

```diff
@@ -0,0 +1,234 @@
+use super::{
+    CallTraceArena, RawOrDecodedCall, RawOrDecodedLog, RawOrDecodedReturnData, TraceIdentifier,
+};
+use crate::abi::{CHEATCODE_ADDRESS, CONSOLE_ABI, HEVM_ABI};
+use ethers::{
+    abi::{Abi, Address, Event, Function, Token},
+    types::H256,
+};
+use foundry_utils::format_token;
+use std::collections::BTreeMap;
+
+/// The call trace decoder.
+///
+/// The decoder collects address labels and ABIs from any number of [TraceIdentifier]s, which it
+/// then uses to decode the call trace.
+///
+/// Note that a call trace decoder is required for each new set of traces, since addresses in
+/// different sets might overlap.
+#[derive(Default, Debug)]
+pub struct CallTraceDecoder {
```

Comment:
The decoder collects information about addresses in the call traces (i.e.. ABIs, labels) using `TraceIdentifier`s, which allows us to pull info from different sources (locally, `vm.label`, 4byte directory, etherscan, ...). It also does the final decoding step for a `CallTraceArena`